pkg/api/types.go

```diff
@@ -518,8 +518,18 @@ type GitRepoVolumeSource struct {
 // The contents of the target Secret's Data field will be presented in a volume
 // as files using the keys in the Data field as the file names.
 type SecretVolumeSource struct {
-	// Name of the secret in the pod's namespace to use
-	SecretName string `json:"secretName"`
+	// Name of the secret in the pod's namespace to use.  May not be specified if items is specified.
+	SecretName string `json:"secretName,omitempty"`
+	// A list of projections to make into the volume.  May not be specified if SecretName is specified.
+	Items []SecretVolumeFile `json:"items,omitempty"`
+}
+
+// SecretVolumeFile represents a single secret key projected into a file.
+type SecretVolumeFile struct {
+	// A reference to the secret key to project into the volume.
+	SecretKeySelector `json:",inline"`
```

Comment:
@thockin 

> How bad would it be for the volume to look like:
>
>    name: MySecrets
>    secret:
>       secretName: app-secret
>       items:
>          - key: username
>            path: username.txt
>          - key: password
>            path: password.txt

I like this better from a stylistic standpoint, actually.  It feels really chatty having to specify the secret name for each file. You are correct, the current state of this PR does make it possible to mount multiple secrets into a volume.  I have no problem with using different structs to implement your API though from a class standpoint it seems a little fine-grained to have to split types, I don't think it's a huge deal.

We wanted the config map and secrets to be similar from an API standpoint, and what you suggest is a change from what we agreed to for config map in https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/configmap.md#consuming-configmap-as-volumes.  So, do we want to consider whether the API for configmap volumes should also change?  From a user standpoint, I guess the question is whether people need to be able to compose a single volume from multiple instances of a resource, or whether volume plugins should deal with a single instance of a resource only.

Thinking about API compatibility, the current state of this PR does seem a little dicey to me now.  An older client would always have to submit valid pods with `secretName` set, but older clients would see a secret volume without `secretName` set, which are invalid.  @bgrant0607 is that an issue?

cc @smarterclayton